VCF writer using numba

tomwhite opened the pull request:
This is a first draft of a `zarr_to_vcf` function that uses numba to get good performance (#924).

- I can get ~100MB/s write speed for the benchmark in https://github.com/pystatgen/sgkit/pull/944 which uses a real VCF file.
- Relies on the VCF header stored as a Zarr attribute, so really only useful for roundtripping VCF -> Zarr -> VCF at the moment.
- It passes some, but not all, of the "lossless" tests that check VCF -> Zarr -> VCF is unchanged.
- Float values are written using the `ftoa` function, which does not have the same output as the htslib implementation in some cases. For example, it has different rules around number of decimal places, and doesn't support exponential notation. This could be improved, but it will be hard to make it match exactly. This is why some of the "lossless" tests are not passing.
-  This is a big PR. The best way to approach it is probably to look at `vcf_writer.py` to get the high-level flow, and `vcf_writer_utils.py` for the smaller numba functions. It would be useful to have a high-level review at least before I do much more on it as a sanity check about some of the design choices and general direction.

hammer commented:
Very cool!

> Float values are written using the ftoa function, which does not have the same output as the htslib implementation in some cases.

Would it be worth doing this work upstream in `cyvcf2` where the work to wrap `htslib` with Cython has been done and see if we can just call out to the `htslib` `ftoa` function?

jeromekelleher commented on sgkit/io/vcf/vcf_writer.py:
```diff
@@ -0,0 +1,291 @@
+
+        k = 0
+        info_prefixes = []  # field names followed by '=' (except for flag/bool types)
+        for key in header_info_fields:
```
Probably worth making ``info_field = f"variant_key"`` here, it's used 3 or 4 times

jeromekelleher commented on sgkit/io/vcf/vcf_writer.py:
```diff
@@ -0,0 +1,291 @@
+        k = 0
+        format_fields = []
+        has_gt = False
+        for key in header_format_fields:
```
Ditto here, could just assign key to variable rather than calling function several times.

jeromekelleher reviewed (commented):
This looks really great @tomwhite. The low-level code paths look just right to me, you've taken a really elegant and clear approach the splitting up the work.

I'm a bit iffy about the top-level interface, but that's something we can iterate on once the basic machinery for doing the writing is merged.

tomwhite commented:
> > Float values are written using the ftoa function, which does not have the same output as the htslib implementation in some cases.
> 
> Would it be worth doing this work upstream in `cyvcf2` where the work to wrap `htslib` with Cython has been done and see if we can just call out to the `htslib` `ftoa` function (however they do it; I did a quick search of their codebase and didn't find a method called `ftoa`).

I'm not sure. The write path currently has no dependency on `cyvcf2` and it doesn't necessarily need to add one, since the VCFs being produced are perfectly valid. We could add it as an option though. On the other hand, the number of decimal places is something that different libraries do differently, e.g. htsjdk [always formats QUAL fields to 2dp](https://github.com/samtools/htsjdk/blob/7719274fe370a51a24e6067de21bbe7e18c160a9/src/main/java/htsjdk/variant/vcf/VCFEncoder.java#L33), but has [more complex rules for other fields](https://github.com/samtools/htsjdk/blob/7719274fe370a51a24e6067de21bbe7e18c160a9/src/main/java/htsjdk/variant/vcf/VCFEncoder.java#L261-L279).

There's also some relationship with the discussion in #925. If a field is using a `FixedScaleOffset`, then we probably want to use that to set the number of decimal places when doing a VCF export.

tomwhite commented:
> This looks really great @tomwhite. The low-level code paths look just right to me, you've taken a really elegant and clear approach the splitting up the work.

Thanks!

> I'm a bit iffy about the top-level interface, but that's something we can iterate on once the basic machinery for doing the writing is merged.

Fair enough - I agree that this part of the API can be improved. I'll have a look at your points by early next week.

jeromekelleher commented:
FWIW I don't think it's worth having a dependency on cyvcf2 for doing the writing - certainly not for the sake of having identical float rounding. Trying to get exact float matching is a fool's game in my experience (rendering binary floats in base 10 is fundamentally lossy), if we're equal to within a small epsilon that's all that matters.

tomwhite commented on sgkit/io/vcf/vcf_writer.py:
```diff
@@ -0,0 +1,291 @@
+
+
+def zarr_to_vcf(
+    input: Union[PathType, MutableMapping[str, bytes]],
```
Added `dataset_to_vcf` and kept `zarr_to_vcf` as a convenience (and counterpart to `vcf_to_zarr`).

tomwhite commented on sgkit/io/vcf/vcf_writer.py:
```diff
@@ -0,0 +1,291 @@
+
+def zarr_to_vcf(
+    input: Union[PathType, MutableMapping[str, bytes]],
+    output: PathType,
```
Fixed

tomwhite commented on sgkit/io/vcf/vcf_writer.py:
```diff
@@ -41,21 +44,23 @@ def zarr_to_vcf(
 
-    with open(output, mode="w") as out:
-        print(header_str, end="", file=out)
+    with ExitStack() as stack:
```
I only just discovered it

tomwhite commented:
I've addressed all the feedback from @jeromekelleher.

The build is failing since the coverage is below 100% since numba-jitted functions are not included (see #77). Normally we add `no cover` pragmas to these functions, but it would be nice to be able to run coverage tests with the jit disabled (#964).

Also, before this is merged I would like to remove the benchmarking code, since this can be done separately (#944).

tomwhite commented:
This is now passing all tests. I've also created some follow up issues (#965, #966, #967).

jeromekelleher reviewed (approved):
Awesome stuff, let's merge!

hammer commented:
Awesome! Should we cut a release to pick up this new functionality?
